kimchi/src/circuits/lookup/constraints.rs

```diff
@@ -47,119 +47,33 @@ pub fn zk_patch<R: Rng + ?Sized, F: FftField>(
 //~ Instead, we arrange the LookupSorted table into columns in a snake-shape.
 //~
 //~ Like so,
+//~
+//~ ```
 //~ _   _
 //~ | | | | |
 //~ | | | | |
 //~ |_| |_| |
+//~ ```
 //~
-//~ or, imagining the full sorted array is [ s0, ..., s8 ], like
+//~ or, imagining the full sorted array is `[ s0, ..., s8 ]`, like
 //~
+//~ ```
 //~ s0 s4 s4 s8
 //~ s1 s3 s5 s7
 //~ s2 s2 s6 s6
+//~ ```
 //~
-//~ So the direction ("increasing" or "decreasing" (relative to LookupTable)
-//~ is
-//~ if i % 2 = 0 { Increasing } else { Decreasing }
+//~ So the direction ("increasing" or "decreasing" (relative to LookupTable) is
 //~
-//~ Then, for each i < max_lookups_per_row, if i % 2 = 0, we enforce that the
-//~ last element of LookupSorted(i) = last element of LookupSorted(i + 1),
-//~ and if i % 2 = 1, we enforce that the
-//~ first element of LookupSorted(i) = first element of LookupSorted(i + 1)
+//~ ```
+//~ if i % 2 = 0 { Increasing } else { Decreasing }
+//~ ```
 //~
-//~ Overview of the protocol
-//~ ========================
-//~ * We have our initial table `lookup_table`, with our desired values listed.
-//~ * We have the implicit table `lookups(witness)` representing the values looked up in each row
-//~   of the witness.
-//~   - This table is initially variable-width, where some rows have no lookups, and others have
-//~     several.
-//~   - We explicitly compute this table, and where the width for a particular row is less than the
-//~     maximum width, we insert a 'dummy' lookup value as many times as we need to to give every
-//~     row the same number of lookups.
-//~   - We'll call this padded table `witness_lookups`.
-//~ * We want to generate a `sorted_table` that contains every entry from the concatenated table
-//~ `lookup_table||witness_lookups`, where values are in the same order as `lookup_table`, with all
-//~ duplicates placed next to each other.
-//~   - There's an edge case around duplicate values in the `lookup_table` itself: these should
-//~     appear in `sorted_table` at least once each time they appeared in the `lookup_table`.
-//~   - This ensures that, for any `beta` and for each `i`, the pair `lookup_table[i] + beta *
-//~     lookup_table[i+1]` corresponds to some distinct `j` such that `sorted_table[j] + beta *
-//~     sorted_table[j+1]`.
-//~   - For all other values of `j`, `sorted_table[j] = sorted_table[j+1]`: since we've dealt with
-//~     all of the 'different' pairs corresponding from moving from one value in `lookup_table` to
-//~     the next, the only remaining pairs are those corresponding to the duplicates provided by the
-//~     lookups in `witness_lookups`.
-//~   - For example, if `lookup_table` is `[0, 1, 2, 3, 4, 5]` and `witness_lookups` is
```

Comment:
It's basically replaced by the RFC https://o1-labs.github.io/proof-systems/rfcs/3-lookup.html

I'm always struggling to find a balance between having this kind of rationale in an RFC or in the spec. It's quite long so I decided to remove it from the code, but I think some of it could be reintroduced in the spec in the lookup section as an overview before we declare the different functions the prover/verifier can use.

Also, it spends quite some time explaining why the grand product argument works, which is not what a spec is for. I think the book would be a better place to talk about intuitions and explain the grand product argument.

wdyt?